systemd/systemd.go

```diff
@@ -400,26 +411,51 @@ type Reporter interface {
 	Notify(string)
 }
 
-func newSystemdReal(be Backend, rootDir string, mode InstanceMode, rep Reporter) Systemd {
+func newSystemdReal(be Backend, rootDir string, mode InstanceMode, rep Reporter, legacy bool) Systemd {
 	switch be {
 	case RunningSystemdBackend:
-		return &systemd{rootDir: rootDir, mode: mode, reporter: rep}
+		return &systemd{rootDir: rootDir, mode: mode, reporter: rep, legacy: legacy}
 	case EmulationModeBackend:
 		return &emulation{rootDir: rootDir}
 	default:
 		panic(fmt.Sprintf("unsupported systemd backend %v", be))
 	}
 }
 
+// MIN_UBUNTU_RELEASE is the Ubuntu release from where we consider systemd being legacy version
+const MIN_UBUNTU_RELEASE = 18
+
+// Based on the release info, determine if systemd should be run in legacy mode
+// or in "smart" mode when we try to avoid daemon-reload calls
+// staying on the safe side, any other than Ubuntu systems are by default in legacy mode
+func legacySystemd() bool {
+	legacy := true
+	// we act smart on Ubuntu Core 20 and Ubuntu 20.04 and newer
+	if "ubuntu-core" == release.ReleaseInfo.ID {
+		ver, err := strconv.Atoi(release.ReleaseInfo.VersionID)
+		if err == nil && ver >= MIN_UBUNTU_RELEASE {
+			legacy = false
+		}
+	} else if "ubuntu" == release.ReleaseInfo.ID {
+		if 2 < len(release.ReleaseInfo.VersionID) {
+			ver, err := strconv.Atoi(release.ReleaseInfo.VersionID[0:2])
+			if err == nil && ver >= MIN_UBUNTU_RELEASE {
+				legacy = false
+			}
+		}
+	}
```

Comment:
I quickly went through a bunch of sytsems we use with spread and all of them support NeedDaemonReload flag which is reported correctly (even 14.04 with systemd 204). Since daemonReloadIfNeededWithLock is the only bit of code that depends on the legacy flag (even in the larger branch #10463) I would suggest we try to use the new mode everywhere and if anything goes wrong in the tests we build a list of systems where we know it does not work.